clang/test/SemaSYCL/intel-fpga-local-ast.cpp

```diff
@@ -0,0 +1,523 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -sycl-std=2020 -ast-dump %s | FileCheck %s
+
+// Tests AST for Intel FPGA memory attributes.
+
+#include "sycl.hpp"
+
+sycl::queue deviceQueue;
+
+//CHECK: FunctionDecl{{.*}}check_ast
+void check_ast()
+{
+  //CHECK: VarDecl{{.*}}doublepump
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelDoublePumpAttr
+  [[intel::doublepump]] unsigned int doublepump[64];
+
+  //CHECK: VarDecl{{.*}}memory
+  //CHECK: SYCLIntelMemoryAttr
+  [[intel::fpga_memory]] unsigned int memory[64];
+
+  //CHECK: VarDecl{{.*}}memory_mlab
+  //CHECK: SYCLIntelMemoryAttr{{.*}}MLAB
+  [[intel::fpga_memory("MLAB")]] unsigned int memory_mlab[64];
+
+  //CHECK: VarDecl{{.*}}mem_blockram
+  //CHECK: SYCLIntelMemoryAttr{{.*}}BlockRAM
+  [[intel::fpga_memory("BLOCK_RAM")]] unsigned int mem_blockram[32];
+
+  //CHECK: VarDecl{{.*}}reg
+  //CHECK: SYCLIntelRegisterAttr
+  [[intel::fpga_register]] unsigned int reg[64];
+
+  //CHECK: VarDecl{{.*}}singlepump
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelSinglePumpAttr
+  [[intel::singlepump]] unsigned int singlepump[64];
+
+  //CHECK: VarDecl{{.*}}bankwidth
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelBankWidthAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}4
+  //CHECK-NEXT: IntegerLiteral{{.*}}4{{$}}
+  [[intel::bankwidth(4)]] unsigned int bankwidth[32];
+
+  //CHECK: VarDecl{{.*}}numbanks
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelNumBanksAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}8
+  //CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
+  [[intel::numbanks(8)]] unsigned int numbanks[32];
+
+  //CHECK: VarDecl{{.*}}private_copies
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelPrivateCopiesAttr
+  //CHECK-NEXT: ConstantExpr
+  //CHECK-NEXT: value:{{.*}}8
+  //CHECK-NEXT: IntegerLiteral{{.*}}8{{$}}
+  [[intel::private_copies(8)]] unsigned int private_copies[64];
+
+  //CHECK: VarDecl{{.*}}merge_depth
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelMergeAttr{{.*}}"mrg1" "depth"{{$}}
+  [[intel::merge("mrg1", "depth")]] unsigned int merge_depth[64];
+
+  //CHECK: VarDecl{{.*}}merge_width
+  //CHECK: SYCLIntelMemoryAttr{{.*}}Implicit
+  //CHECK: SYCLIntelMergeAttr{{.*}}"mrg2" "width"{{$}}
+  [[intel::merge("mrg2", "width")]] unsigned int merge_width[64];
+
+  //CHECK: VarDecl{{.*}}bankbits
+  //CHECK: SYCLIntelNumBanksAttr{{.*}}Implicit{{$}}
+  //CHECK-NEXT: IntegerLiteral{{.*}}16{{$}}
```

Comment:
What is this value 16 that is being checked?

Reply:
Thanks @premanandrao for reviews.

```
/// The numbanks and bank_bits attributes are related. If numbanks exists
/// when handling bank_bits they are checked for consistency. If numbanks
/// hasn't been added yet an implicit one is added with the correct value.
/// If the user later adds a numbanks attribute the implicit one is removed.
/// The values must be consecutive values (i.e. 3,4,5 or 2,1).
````

Reply:
Thanks for explaining.   Can you add a test or modify one where the user-specified numbanks value overrides the value computed from the bank_bits arguments?

Reply:
@premanandrao i have added test. 

User-specified numbank attribute overrides implict numbank attribute computed from bank_bits attributes. But value should be same since the number of bank_bits must be equal to ceil(log2(numbanks))
```
//expected-error@+1{{the number of bank_bits must be equal to ceil(log2(numbanks))}}
[[intel::numbanks(8), intel::bank_bits(3, 4)]] unsigned int bb_numbanks[4];
```

```
DeclStmt 0x5596d08cb8f0 <line:106:3, col:84>
|   | `-VarDecl 0x5596d08cb6f0 <col:52, col:83> col:65 bank_bits_width 'unsigned int[64]'
|   |   |-SYCLIntelNumBanksAttr 0x5596d08cb7d0 <<invalid sloc>> Implicit
|   |   | `-IntegerLiteral 0x5596d08cb7b0 <<invalid sloc>> 'int' 4
|   |   |-SYCLIntelMemoryAttr 0x5596d08cb830 <<invalid sloc>> Implicit Default
|   |   |-SYCLIntelBankBitsAttr 0x5596d08cb858 <col:5, col:26>
|   |   | |-ConstantExpr 0x5596d08cb770 <col:22> 'int'
|   |   | | |-value: Int 2
|   |   | | `-IntegerLiteral 0x5596d08cb628 <col:22> 'int' 2
|   |   | `-ConstantExpr 0x5596d08cb790 <col:25> 'int'
|   |   |   |-value: Int 3
|   |   |   `-IntegerLiteral 0x5596d08cb648 <col:25> 'int' 3
|   |   `-SYCLIntelBankWidthAttr 0x5596d08cb8c0 <col:29, col:48>
|   |     `-ConstantExpr 0x5596d08cb8a0 <col:46> 'int'
|   |       |-value: Int 16
|   |       `-IntegerLiteral 0x5596d08cb668 <col:46> 'int' 16

//User specified SYCLIntelNumBanksAttr.

|   |-DeclStmt 0x5596d08cbbd0 <line:123:3, col:80>
|   | `-VarDecl 0x5596d08cb9d0 <col:50, col:79> col:63 bank_bits_num 'unsigned int[64]'
|   |   |-SYCLIntelMemoryAttr 0x5596d08cbb10 <<invalid sloc>> Implicit Default
|   |   |-SYCLIntelBankBitsAttr 0x5596d08cbb38 <col:5, col:26>
|   |   | |-ConstantExpr 0x5596d08cba50 <col:22> 'int'
|   |   | | |-value: Int 2
|   |   | | `-IntegerLiteral 0x5596d08cb908 <col:22> 'int' 2
|   |   | `-ConstantExpr 0x5596d08cba70 <col:25> 'int'
|   |   |   |-value: Int 3
|   |   |   `-IntegerLiteral 0x5596d08cb928 <col:25> 'int' 3
|   |   `-SYCLIntelNumBanksAttr 0x5596d08cbba0 <col:29, col:46>
|   |     `-ConstantExpr 0x5596d08cbb80 <col:45> 'int'
|   |       |-value: Int 4
|   |       `-IntegerLiteral 0x5596d08cb948 <col:45> 'int' 4

````